src/payload/addr.rs

```diff
@@ -0,0 +1,711 @@
+    }
+}
+
+
+//--- FromStr and Display
+
+impl FromStr for Prefix {
+    type Err = ParsePrefixError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.is_empty() {
+            return Err(ParsePrefixError::Empty)
+        }
+        let slash = s.find('/').ok_or(ParsePrefixError::MissingLen)?;
+        let addr = IpAddr::from_str(&s[..slash]).map_err(
+            ParsePrefixError::InvalidAddr
+        )?;
+        let len = u8::from_str(&s[slash + 1..]).map_err(
+            ParsePrefixError::InvalidLen
+        )?;
+        Prefix::new(addr, len).map_err(ParsePrefixError::InvalidPrefix)
+    }
+}
+
+impl fmt::Display for Prefix {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}/{}", self.addr(), self.len())
+    }
+}
+
+
+
+//------------ MaxLenPrefix --------------------------------------------------
+
+/// The pair of a prefix and an optional max-len.
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct MaxLenPrefix {
+    /// The prefix.
+    prefix: Prefix,
+
+    /// The optional maximum prefix length.
+    max_len: Option<u8>,
+}
+
+impl MaxLenPrefix {
+    /// Creates a new value.
+    ///
+    /// The function returns an error if `max_len` is present and smaller than
+    /// `prefix.len()` or larger than the maximum prefix length of the
+    /// prefix’s address family.
+    pub fn new(
+        prefix: Prefix, max_len: Option<u8>
+    ) -> Result<Self, MaxLenError> {
+        if let Some(max_len) = max_len {
+            if
+                (prefix.is_v4() && max_len > 32)
+                || max_len > 128
+            {
+                return Err(MaxLenError::Overflow)
+            }
+            if prefix.len() > max_len {
+                return Err(MaxLenError::Underflow)
+            }
+        }
+        Ok(MaxLenPrefix { prefix, max_len })
+    }
+
+    /// Creates a value curtailing any out-of-bounds max-len.
+    pub fn saturating_new(prefix: Prefix, max_len: Option<u8>) -> Self {
+        let max_len = max_len.map(|max_len| {
+            if prefix.len() > max_len {
+                prefix.len()
+            }
+            else if prefix.is_v4() && max_len > 32 {
+                32
+            }
+            else if max_len > 128 {
+                128
+            }
+            else {
+                max_len
+            }
+        });
+        MaxLenPrefix { prefix, max_len }
+    }
+
+    /// Returns the actual prefix.
+    pub fn prefix(self) -> Prefix {
+        self.prefix
+    }
+
+    /// Returns the address of the prefix.
+    pub fn addr(self) -> IpAddr {
+        self.prefix.addr()
+    }
+
+    /// Returns the prefix length.
+    pub fn prefix_len(self) -> u8 {
+        self.prefix.len()
+    }
+
+    /// Returns the max-length.
+    pub fn max_len(self) -> Option<u8> {
+        self.max_len
+    }
+
+    /// Returns the max-length or the prefix-length if there is no max-length.
+    pub fn unwrapped_max_len(self) -> u8 {
```

Comment:
‘Unwrap’ is probably the wrong term. I want to keep `max_len` returning the option, though (because the actual max len is optional), so I’ll switch this function to be `resolved_max_len`.